File: src/main/java/org/hisp/dhis/rules/models/Rule.java
package org.hisp.dhis.rules.models;

import javax.annotation.CheckForNull;
import javax.annotation.Nonnull;
import java.util.List;

;

public record Rule(
    @CheckForNull
    String name,
    @CheckForNull
    String programStage,
    @CheckForNull
    Integer priority,
    @Nonnull
    String condition,
    @Nonnull
    List<RuleAction> actions,
    @Nonnull
    String uid) {

    public static final Rule MOCK = new Rule(null, null, null, "true", List.of(), null);

    public static Rule copy( @Nonnull Rule rule, @Nonnull List<RuleAction> actions )
    {
        return new Rule(rule.name(), rule.programStage(), rule.priority(), rule.condition(),
                List.copyOf(actions), rule.uid());
    }

    @Nonnull
    public static Rule create( @CheckForNull String programStage, @CheckForNull Integer priority,
        @Nonnull String condition, @Nonnull List<RuleAction> actions, @CheckForNull String name, @Nonnull String uid )
    {
        return new Rule(name, programStage, priority, condition, List.copyOf(actions), uid);
    }


}

File: src/main/java/org/hisp/dhis/rules/models/Rule.kt
package org.hisp.dhis.rules.models

data class Rule(
    val condition: String,
    val actions: List<RuleAction>,
    val uid: String = "",
    val name: String? = null,
    val programStage: String? = null,
    val priority: Int? = 0
){

    fun name(): String? {
        return name
    }

    fun condition(): String {
        return condition
    }

    fun actions(): List<RuleAction> {
        return actions
    }

    fun uid(): String {
        return uid
    }

    fun programStage(): String? {
        return programStage
    }

    fun priority(): Int? {
        return priority
    }
}
